### ISSUE TITLE
Reproduce Gimp Convolution in jimp (Convolution matrix)

### ISSUE THREAD

**Issue by operationexodus** [oliver-moran/jimp#368]
I can not reproduce the same result I have on Gimp. What I do wrong?

Original Image:
![photo_2017-11-28_12-46-00](https://user-images.githubusercontent.com/11057215/33370652-810efec4-d4f8-11e7-8037-7c789b0c3134.jpg)

With gimp I use this procedure: Filter-> Generic-> Convolution Matrix. 
Settings: 
![cattura](https://user-images.githubusercontent.com/11057215/33370887-382ca138-d4f9-11e7-9036-ee1ce0671ee4.PNG)
Result: 
![photo_2017-11-28_12-46-00_d](https://user-images.githubusercontent.com/11057215/33370933-6400c8d4-d4f9-11e7-81bb-b37f8f188125.jpg)

With jimp "0.2.28" on Node:
```javascript
original_image = original_image.convolute([
            [0, 0, 0, 0, 0],
            [0, 1, 1, 1, 0],
            [0, 1, 0, 1, 0],
            [0, 1, 1, 1, 0],
            [0, 0, 0, 0, 0],
        ]);
```
Result:
![1](https://user-images.githubusercontent.com/11057215/33371060-cb02f8ea-d4f9-11e7-83f4-5b5d520a264f.jpeg)

Thanks!

**Comment by andq-ibl** [comment]
The problem is that the convolution function in Jimp does not handle the case that new pixel value is greater than 255

**Comment by steffkelsey** [comment]
You can try monkey-patching the convolute function to this:
```Jimp.prototype.convolute = function (kernel, x, y, w, h, cb) {
    if (!Array.isArray(kernel))
        return throwError.call(this, "the kernel must be an array", cb);

    if (typeof x === "function") {
        cb = x;
        x = y = w = h = null;
    } else {
        if (isDef(x) && typeof x !== "number")
            return throwError.call(this, "x must be a number", cb);
        if (isDef(y) && typeof y !== "number")
            return throwError.call(this, "y must be a number", cb);
        if (isDef(w) && typeof w !== "number")
            return throwError.call(this, "w must be a number", cb);
        if (isDef(h) && typeof h !== "number")
            return throwError.call(this, "h must be a number", cb);
    }

    let ksize = (kernel.length - 1) / 2;

    x = isDef(x) ? x : ksize;
    y = isDef(y) ? y : ksize;
    w = isDef(w) ? w : this.bitmap.width - x;
    h = isDef(h) ? h : this.bitmap.height - y;

    var source = this.cloneQuiet();
    this.scanQuiet(x, y, w, h, function (xx, yx, idx) {
        var value = applyKernel(source, kernel, xx, yx);

        this.bitmap.data[idx] = Math.max(0, Math.min(value[0], 255));
        this.bitmap.data[idx + 1] = Math.max(0, Math.min(value[1], 255));
        this.bitmap.data[idx + 2] = Math.max(0, Math.min(value[2], 255));
    });

    if (isNodePattern(cb)) return cb.call(this, null, this);
    else return this;
};```

That would keep all the final values between 0 and 255.

**Comment by hipstersmoothie** [comment]
@steffkelsey why monkey patch when you can make a PR. please do!

**Comment by hipstersmoothie** [comment]
Released in [v0.4.0](https://github.com/oliver-moran/jimp/releases/tag/v0.4.0)
